offload/plugins-nextgen/level_zero/include/L0Device.h

```diff
@@ -0,0 +1,643 @@
+  }
+  uint32_t getNumSlices() const { return DeviceProperties.numSlices; }
+  uint32_t getNumSubslices() const {
+    return DeviceProperties.numSubslicesPerSlice * DeviceProperties.numSlices;
+  }
+  uint32_t getNumEUs() const {
+    return DeviceProperties.numEUsPerSubslice * getNumSubslices();
+  }
+  uint32_t getTotalThreads() const {
+    return DeviceProperties.numThreadsPerEU * getNumEUs();
+  }
+  uint32_t getNumThreadsPerSubslice() const {
+    return getNumEUsPerSubslice() * getNumThreadsPerEU();
+  }
+  uint32_t getClockRate() const { return DeviceProperties.coreClockRate; }
+
+  uint32_t getMaxSharedLocalMemory() const {
+    return ComputeProperties.maxSharedLocalMemory;
+  }
+  uint32_t getMaxGroupSize() const {
+    return ComputeProperties.maxTotalGroupSize;
+  }
+  uint32_t getMaxGroupCount() const {
+    return getMaxGroupCountX() * getMaxGroupCountY() * getMaxGroupCountZ();
+  }
+
+  uint32_t getMaxGroupSizeX() const { return ComputeProperties.maxGroupSizeX; }
+  uint32_t getMaxGroupSizeY() const { return ComputeProperties.maxGroupSizeY; }
+  uint32_t getMaxGroupSizeZ() const { return ComputeProperties.maxGroupSizeZ; }
+  uint32_t getMaxGroupCountX() const {
+    return ComputeProperties.maxGroupCountX;
+  }
+  uint32_t getMaxGroupCountY() const {
+    return ComputeProperties.maxGroupCountY;
+  }
+  uint32_t getMaxGroupCountZ() const {
+    return ComputeProperties.maxGroupCountZ;
+  }
+  uint32_t getMemoryClockRate() const { return MemoryProperties.maxClockRate; }
+  uint64_t getGlobalMemorySize() const { return MemoryProperties.totalSize; }
+  size_t getCacheSize() const { return CacheProperties.cacheSize; }
+  uint64_t getMaxMemAllocSize() const {
+    return DeviceProperties.maxMemAllocSize;
+  }
+
+  int32_t getAllocKind() const { return AllocKind; }
+  DeviceArchTy getDeviceArch() const { return DeviceArch; }
+  bool isDeviceArch(DeviceArchTy Arch) const { return DeviceArch == Arch; }
+
+  static bool isDiscrete(uint32_t PCIId) {
+    switch (static_cast<PCIIdTy>(PCIId & 0xFF00)) {
+    case PCIIdTy::BMG:
+      return true;
+    default:
+      return false;
+    }
+  }
+
+  static bool isDiscrete(ze_device_handle_t Device) {
+    ze_device_properties_t PR{};
+    PR.stype = ZE_STRUCTURE_TYPE_DEVICE_PROPERTIES;
+    PR.pNext = nullptr;
+    CALL_ZE_RET(false, zeDeviceGetProperties, Device, &PR);
+    return isDiscrete(PR.deviceId);
+  }
+
+  bool isDiscreteDevice() { return isDiscrete(getPCIId()); }
+  bool isDeviceIPorNewer(uint32_t Version) const;
+
+  const std::string_view getUuid() const { return DeviceUuid; }
+
+  uint32_t getComputeEngine() const { return ComputeOrdinal.first; }
+  uint32_t getNumComputeQueues() const { return ComputeOrdinal.second; }
+
+  bool hasMainCopyEngine() const { return CopyOrdinal.first != MaxOrdinal; }
+  uint32_t getMainCopyEngine() const { return CopyOrdinal.first; }
+
+  bool deviceRequiresImmCmdList() const {
+    return isDeviceIPorNewer(0x05004000);
```

Comment:
nit: maybe we use a `constexpr` local variable to explain what the magic number argument is?